readthedocs/oauth/services/gitlab.py

```diff
@@ -0,0 +1,222 @@
+                avatar=fields['avatar']['url'],
+            )
+        organization.json = json.dumps(fields)
+        organization.save()
+        return organization
+
+    def setup_webhook(self, project):
+        """Set up GitLab project webhook for project
+
+        :param project: project to set up webhook for
+        :type project: Project
+        :returns: boolean based on webhook set up success
+        :rtype: bool
+        """
+        session = self.get_session()
+
+        # See: http://doc.gitlab.com/ce/api/projects.html#add-project-hook
+        data = json.dumps({
+            'id': 'readthedocs',
+            'push_events': True,
+            'issues_events': False,
+            'merge_requests_events': False,
+            'note_events': False,
+            'tag_push_events': True,
+            'url': 'https://{0}/gitlab'.format(settings.PRODUCTION_DOMAIN),
+        })
+        resp = None
+        try:
+            repositories = RemoteRepository.objects.filter(
+                clone_url=project.vcs_repo().repo_url
+            )
+            assert repositories
+            repo_id = repositories[0].get_serialized()['id']
+            resp = session.post(
+                '{url}/api/v3/projects/{repo_id}/hooks'.format(
+                    url=self.adapter.provider_base_url,
+                    repo_id=repo_id,
+                ),
+                data=data,
+                headers={'content-type': 'application/json'}
+            )
+            if resp.status_code == 201:
+                log.info('GitLab webhook creation successful for project: %s',  # noqa
+                         project)
+                return True
+        except (AssertionError,  RemoteRepository.DoesNotExist) as ex:
+            log.error('GitLab remote repository not found', exc_info=ex)
+        except RequestException as ex:
+            pass
+        else:
+            ex = False
+
+        log.error('GitLab webhook creation failed for project: %s',  # noqa
+                  project, exc_info=ex)
+
+    @classmethod
+    def get_token_for_project(cls, project, force_local=False):
+        """Get access token for project by iterating over project users"""
+        # TODO why does this only target GitHub?
+        if not getattr(settings, 'ALLOW_PRIVATE_REPOS', False):
+            return None
+        token = None
+        try:
+            if getattr(settings, 'DONT_HIT_DB', True) and not force_local:
+                token = api.project(project.pk).token().get()['token']
+            else:
+                for user in project.users.all():
+                    tokens = SocialToken.objects.filter(
+                        account__user=user,
+                        app__provider=cls.adapter.provider_id)
+                    if tokens.exists():
+                        token = tokens[0].token
+        except Exception:
+            log.error('Failed to get token for user', exc_info=True)
+        return token
```

Comment:
Not entirely sure this is required, or it could be moved to the base class. https://github.com/rtfd/readthedocs.org/blob/e4958838a512b095d0bc8cdf8617c30cc9e489d4/readthedocs/restapi/views/model_views.py#L83-L90 was only targeting Github before refactoring oauth services.